src/plugins/setup.ts

```diff
@@ -0,0 +1,121 @@
+import { PluginAttachment } from '@posthog/plugin-scaffold'
+
+import { getPluginAttachmentRows, getPluginConfigRows, getPluginRows } from '../sql'
+import { status } from '../status'
+import { Plugin, PluginConfig, PluginConfigId, PluginId, PluginsServer, TeamId } from '../types'
+import { loadPlugin } from './loadPlugin'
+
+export async function setupPlugins(server: PluginsServer): Promise<void> {
+    const { plugins, pluginConfigs, pluginConfigsPerTeam } = await loadPluginsFromDB(server)
+
+    for (const [id, pluginConfig] of pluginConfigs) {
+        const plugin = plugins.get(pluginConfig.plugin_id)
+        const prevConfig = server.pluginConfigs.get(id)
+        const prevPlugin = prevConfig ? server.plugins.get(pluginConfig.plugin_id) : null
+
+        // :TRICKY: This forces a reload for plugin VMs which have either been added or changed
+        if (
+            prevConfig &&
+            pluginConfig.updated_at === prevConfig.updated_at &&
+            plugin?.updated_at == prevPlugin?.updated_at
+        ) {
+            pluginConfig.vm = prevConfig.vm
+        }
+
+        if (!pluginConfig.vm) {
+            await loadPlugin(server, pluginConfig)
+        }
```

Comment:
Possible optimisation: push all the promises into an array and `await Promise.all(loadPluginPromises)`. This should speed up loading even more. 

One future danger: we might one day load thousands of plugins in parallel this way. For that there's a function `groupIntoBatches` in `utils.ts` that could split the array into chunks of 100, each of which can be awaited sequentially...

Reply:
Good note. I _think_ no chunking is needed here since node.js only does things in event loop and we're not doing _too_ much async in there - all the file reads are sync and we're only hitting the db minimally.